#!/bin/sh -e
## 670_patch_desireddf_bug402171 by Giuseppe Sacco <eppesuig@debian.org>
##
## All lines beginning with \`## DP:' are a description of the patch.
## DP: Patch for Debian bug #402171 as published for HylaFAX+ on
## DP: http://article.gmane.org/gmane.comp.telephony.fax.hylafax.user/21188

if [ $# -lt 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch} ${2:+-d $2}"

case "$1" in
       -patch) patch $patch_opts -p1 < $0;;
       -unpatch) patch $patch_opts -p1 -R < $0;;
        *)
                echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
                exit 1;;
esac

exit 0

#@DPATCH@
#EOF
diff -urNad hylafax-4.3.1~/faxd/FaxSend.c++ hylafax-4.3.1/faxd/FaxSend.c++
--- hylafax-4.3.1~/faxd/FaxSend.c++	2006-04-28 22:05:57.000000000 +0200
+++ hylafax-4.3.1/faxd/FaxSend.c++	2006-12-09 18:53:20.000000000 +0100
@@ -43,8 +43,9 @@
  * FAX Server Transmission Protocol.
  */
 void
-FaxServer::sendFax(FaxRequest& fax, FaxMachineInfo& clientInfo, FaxAcctInfo& ai, u_int& batched)
+FaxServer::sendFax(FaxRequest& fax, FaxMachineInfo& clientInfo, FaxAcctInfo& ai, u_int& batched, bool usedf)
 {
+    useDF = usedf;
     u_int prevPages = fax.npages;
     if (!(batched & BATCH_FIRST) || lockModem()) {
         if (batched & BATCH_FIRST)
@@ -163,7 +164,7 @@
      */
     clientParams.decodePage(fax.pagehandling);
     /*
-     * So we want to restrict DF sellection to those masked in fax.desireddf
+     * So we want to restrict DF selection to those masked in fax.desireddf
      */
     clientParams.df = fxmin(modem->getBestDataFormat(), (u_int)fax.desireddf);
     clientParams.br = fxmin(modem->getBestSignallingRate(), (u_int) fax.desiredbr);
@@ -606,12 +607,21 @@
 	 * Ignore what faxq did and send with the "highest" (monochrome) 
 	 * compression that both the modem and the remote supports.
 	 */
-	params.df = 0;
-	u_int bits = clientCapabilities.df;
-	bits &= BIT(DF_JBIG+1)-1;		// cap at JBIG, only deal with monochrome
-	while (bits) {
-	    bits >>= 1;
-	    if (bits) params.df++;
+	if (useDF) {
+	    /*
+	     * The job has been restricted to a specific data format per
+	     * JobControls or something similar - that value is now 
+	     * found in clientParams.df.
+	     */
+	    params.df ==  fxmin(clientParams.df, params.df);
+	} else {
+	    params.df = 0;
+	    u_int bits = clientCapabilities.df;
+	    bits &= BIT(DF_JBIG+1)-1;		// cap at JBIG, only deal with monochrome
+	    while (bits) {			// puts params.df to the "best" support by the remote
+		bits >>= 1;
+		if (bits) params.df++;
+	    }
 	}
 	if (params.df == DF_JBIG && (!modem->supportsJBIG() || (params.ec == EC_DISABLE)))
 		params.df = DF_2DMMR;
@@ -619,7 +629,7 @@
 	// it's likely that the remote was incorrect in telling us it does
 	if (params.df == DF_2DMRUNCOMP) params.df = DF_2DMR;
 	// don't let RTFCC cause problems with restricted modems...
-	if (params.df == DF_2DMMR && (!modem->supportsMMR() || (params.ec == EC_DISABLE)))
+	if (params.df == DF_2DMMR && (!modem->supportsMMR() || (params.ec == EC_DISABLE) || !(clientCapabilities.df & BIT(DF_2DMMR))))
 		params.df = DF_2DMR;
 	if (params.df == DF_2DMR && (!modem->supports2D() || !(clientCapabilities.df & BIT(DF_2DMR))))
 		params.df = DF_1DMH;
diff -urNad hylafax-4.3.1~/faxd/FaxServer.h hylafax-4.3.1/faxd/FaxServer.h
--- hylafax-4.3.1~/faxd/FaxServer.h	2006-03-17 17:47:09.000000000 +0100
+++ hylafax-4.3.1/faxd/FaxServer.h	2006-12-09 18:54:25.000000000 +0100
@@ -63,6 +63,8 @@
 
     friend class FaxModem;
 
+    bool	useDF;			// limits application of RTFCC
+
 // FAX transmission protocol support
     void	sendFax(FaxRequest& fax, FaxMachineInfo&, const fxStr& number, u_int&);
     bool	sendClientCapabilitiesOK(FaxRequest&, FaxMachineInfo&, fxStr&);
@@ -93,7 +95,7 @@
     void	readConfig(const fxStr& filename);
     void	setLocalIdentifier(const fxStr& lid);
 
-    void	sendFax(FaxRequest&, FaxMachineInfo&, FaxAcctInfo&, u_int&);
+    void	sendFax(FaxRequest&, FaxMachineInfo&, FaxAcctInfo&, u_int&, bool);
     bool	recvFax(const CallID& callid, fxStr& emsg);
 
     time_t	getFileTransferTime() const;
diff -urNad hylafax-4.3.1~/faxd/JobControl.c++ hylafax-4.3.1/faxd/JobControl.c++
--- hylafax-4.3.1~/faxd/JobControl.c++	2006-03-28 15:22:51.000000000 +0200
+++ hylafax-4.3.1/faxd/JobControl.c++	2006-12-09 18:57:45.000000000 +0100
@@ -38,6 +38,7 @@
 #define	DCI_MAXTRIES		0x0010
 #define	DCI_USEXVRES		0x0020
 #define	DCI_VRES		0x0040
+#define	DCI_DESIREDDF		0x0100
 
 #define	isDefined(b)		(defined & b)
 #define	setDefined(b)		(defined |= b)
@@ -56,6 +57,7 @@
     maxTries = other.maxTries;
     usexvres = other.usexvres;
     vres = other.vres;
+    desireddf = other.desireddf;
 }
 
 JobControlInfo::JobControlInfo (const fxStr& buffer)
@@ -131,10 +133,14 @@
 	vres = getNumber(value);
 	setDefined(DCI_VRES);
     } else {
+	if (streq(tag, "desireddf")) {		// need to pass desireddf to faxsend, also
+	    desireddf = getNumber(value);
+	    setDefined(DCI_DESIREDDF);
+	}
 	if( args != "" )
 	    args.append('\0');
 	args.append(fxStr::format("-c%c%s:\"%s\"",
-	    '\0', tag, value));
+	    '\0', tag, (const char*) value));
     }
     return true;
 }
@@ -213,3 +219,12 @@
     else
 	return 0;
 }
+
+int
+JobControlInfo::getDesiredDF() const
+{
+    if (isDefined(DCI_DESIREDDF))
+	return desireddf;
+    else
+	return -1;
+}
diff -urNad hylafax-4.3.1~/faxd/JobControl.h hylafax-4.3.1/faxd/JobControl.h
--- hylafax-4.3.1~/faxd/JobControl.h	2006-03-28 15:22:51.000000000 +0200
+++ hylafax-4.3.1/faxd/JobControl.h	2006-12-09 18:59:11.000000000 +0100
@@ -54,6 +54,7 @@
     int		usexvres;		// use extended resolution
     u_int	vres;			// use extended resolution
     fxStr	args;			// arguments for subprocesses
+    int		desireddf;		// if set, desireddf value
 
     // default returned on no match
     static const JobControlInfo defControlInfo;
@@ -66,7 +67,6 @@
     ~JobControlInfo();
 
     int compare(const JobControlInfo*) const;
-    void parseEntry(const char* tag, const char* value, bool quoted);
 
     u_int getMaxConcurrentCalls() const;
     u_int getMaxSendPages() const;
@@ -77,6 +77,7 @@
     time_t nextTimeToSend(time_t) const;
     int getUseXVRes() const;
     u_int getVRes() const;
+    int getDesiredDF() const;
     const fxStr& getArgs() const;
 
     virtual bool setConfigItem(const char*, const char*);
diff -urNad hylafax-4.3.1~/faxd/faxQueueApp.c++ hylafax-4.3.1/faxd/faxQueueApp.c++
--- hylafax-4.3.1~/faxd/faxQueueApp.c++	2006-10-20 07:29:16.000000000 +0200
+++ hylafax-4.3.1/faxd/faxQueueApp.c++	2006-12-09 18:48:29.000000000 +0100
@@ -660,6 +660,8 @@
      * o the remote side is known to be capable of it, and
      * o the user hasn't specified a desire to send 1D data.
      */
+    int jcdf = job.getJCI().getDesiredDF();
+    if (jcdf != -1) req.desireddf = jcdf;
     if (req.desireddf == DF_2DMMR && (req.desiredec != EC_DISABLE) && 
 	use2D && job.modem->supportsMMR() &&
 	 (! info.getCalledBefore() || info.getSupportsMMR()) )
diff -urNad hylafax-4.3.1~/faxd/faxSendApp.c++ hylafax-4.3.1/faxd/faxSendApp.c++
--- hylafax-4.3.1~/faxd/faxSendApp.c++	2006-03-28 01:00:54.000000000 +0200
+++ hylafax-4.3.1/faxd/faxSendApp.c++	2006-12-09 18:50:16.000000000 +0100
@@ -161,8 +161,11 @@
 			 * user-specification.
 			 */
 
-			if (desiredDF != (u_int) -1)
+			bool usedf = false;			// to limit RTFCC application
+			if (desiredDF != (u_int) -1) {
 			    req->desireddf = desiredDF;
+			    usedf = true;
+			}
 			if (desiredBR != (u_int) -1)
 			    req->desiredbr = desiredBR;
 			if (desiredEC != (u_int) -1)
@@ -175,7 +178,7 @@
 			if (useJobTSI && req->tsi != "")
 			    FaxServer::setLocalIdentifier(req->tsi);
 
-			FaxServer::sendFax(*req, info, ai, batched);
+			FaxServer::sendFax(*req, info, ai, batched, usedf);
 
 			batchcommid = req->commid;		// ... to all batched jobs
 
diff -urNad hylafax-4.3.1~/man/jobcontrol.1m hylafax-4.3.1/man/jobcontrol.1m
--- hylafax-4.3.1~/man/jobcontrol.1m	2006-03-28 01:00:57.000000000 +0200
+++ hylafax-4.3.1/man/jobcontrol.1m	2006-12-09 18:59:35.000000000 +0100
@@ -164,6 +164,7 @@
 .ft C
     DesiredBR: 3
     DesiredEC: 0
+    DesiredDF: 1
 .ft P
 .fi
 .SH NOTES
